// @anna-ai/cli/dist/apps-sync-meta-BOb6J_wz.js
import { canonicalHost } from "./credentials-BTv2IfUZ.js";
import { findAppBySlug, getApp } from "./apps-DlVsVHQ5.js";
import { CliError } from "./client-D-_z1ALk.js";
import { resolveClient, withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import { loadAppManifest } from "./manifest-DOMrcdCK.js";
import { readAppIdentity } from "./app-cache-BqnTkxUw.js";
import { syncAppListingMeta } from "./listing-meta-BlY7XMrw.js";
import { resolve } from "node:path";
import { bold, cyan, dim, green } from "kleur/colors";

//#region src/commands/apps-sync-meta.ts
async function runAppsSyncMeta(opts) {
	return withErrorHandling(async () => {
		const cwd = resolve(opts.cwd ?? process.cwd());
		const manifest = loadAppManifest(cwd, opts.manifest);
		const { client, host } = resolveClient({ account: opts.account });
		let appId = null;
		let serverSlug = manifest.slug;
		const cached = readAppIdentity(cwd);
		if (cached && cached.host === canonicalHost(host) && cached.slug === manifest.slug) {
			appId = cached.app_id;
			serverSlug = cached.slug;
		} else {
			const found = await findAppBySlug(client, manifest.slug);
			if (found) {
				appId = found.id;
				serverSlug = found.slug;
			}
		}
		if (appId === null) throw new CliError(`no app with slug "${manifest.slug}" on ${host} — run 'anna-app apps publish' first`, 5);
		const sync = await syncAppListingMeta(client, appId, manifest, {
			cwd,
			logoOverride: opts.logo,
			dryRun: opts.dryRun
		});
		const meta = sync.meta;
		if (opts.dryRun) {
			const plan = {
				host,
				app_id: appId,
				slug: serverSlug,
				meta,
				logo: sync.logoUpload?.source,
				screenshots: sync.screenshotUploads?.map((s) => s.source),
				dry_run: true
			};
			if (opts.json) {
				console.log(JSON.stringify(plan, null, 2));
				return 0;
			}
			console.log(`[dry-run] would PATCH apps/${serverSlug} metadata: ` + JSON.stringify(meta));
			if (sync.logoUpload) console.log(`[dry-run] would upload logo from ${sync.logoUpload.source}`);
			for (const shot of sync.screenshotUploads ?? []) console.log(`[dry-run] would upload screenshot from ${shot.source}`);
			return 0;
		}
		const finalApp = await getApp(client, appId);
		if (opts.json) {
			console.log(JSON.stringify({
				host,
				app_id: appId,
				slug: serverSlug,
				meta,
				logo_url: sync.logoUpload?.logo_url,
				screenshot_urls: sync.screenshotUploads?.map((s) => s.screenshot_url),
				status: finalApp.status
			}, null, 2));
			return 0;
		}
		console.log(green(`✓ apps/${serverSlug}: metadata synced`));
		console.log(`  ${dim("name     :")} ${manifest.name}`);
		if (manifest.category) console.log(`  ${dim("category :")} ${manifest.category}`);
		if (manifest.tagline) console.log(`  ${dim("tagline  :")} ${manifest.tagline}`);
		if (sync.logoUpload?.logo_url) console.log(`  ${dim("logo     :")} ${sync.logoUpload.logo_url}`);
		if (sync.screenshotUploads && sync.screenshotUploads.length > 0) console.log(`  ${dim("shots    :")} ${sync.screenshotUploads.length} uploaded`);
		console.log(bold(cyan(`status: ${finalApp.status}`)));
		return 0;
	});
}

//#endregion
export { runAppsSyncMeta };
