// @anna-ai/cli/dist/listing-meta-BlY7XMrw.js
import { patchApp, uploadAppLogo, uploadAppScreenshot } from "./apps-DlVsVHQ5.js";
import { CliError } from "./client-D-_z1ALk.js";
import { extname, resolve } from "node:path";
import { existsSync, readFileSync } from "node:fs";

//#region src/publish/listing-meta.ts
const IMAGE_CONTENT_TYPES = {
	".svg": "image/svg+xml",
	".png": "image/png",
	".jpg": "image/jpeg",
	".jpeg": "image/jpeg",
	".webp": "image/webp",
	".gif": "image/gif"
};
function imageContentType(path) {
	return IMAGE_CONTENT_TYPES[extname(path).toLowerCase()] ?? "application/octet-stream";
}
/**
* A listing asset is "remote" (forwarded verbatim) when it's an absolute
* http(s) URL or an inline `data:` URI. Anything else is treated as a
* project-relative local file to be uploaded to the CDN — same split the logo
* uses between `logoUrl` and `logoFile`.
*/
function isRemoteAsset(src) {
	return /^https?:\/\//i.test(src) || src.startsWith("data:");
}
/**
* Build the metadata payload from the manifest (only defined fields). The
* logo is handled separately and never appears here when sourced from a file.
*/
function buildMeta(manifest) {
	const meta = {};
	const add = (key, value) => {
		if (value !== void 0 && value !== null && value !== "") meta[key] = value;
	};
	add("name", manifest.name);
	add("category", manifest.category);
	add("tagline", manifest.tagline);
	add("description", manifest.description);
	add("cover_url", manifest.coverUrl);
	add("homepage_url", manifest.homepageUrl);
	add("support_url", manifest.supportUrl);
	add("privacy_url", manifest.privacyUrl);
	add("pricing_model", manifest.pricingModel);
	return meta;
}
/**
* Sync the app's store-listing metadata (and logo) from the local manifest.
*
* Returns a plan/result. When `dryRun` is set, no network calls are made and
* the returned object reflects what *would* be sent.
*/
async function syncAppListingMeta(client, appId, manifest, opts) {
	const meta = buildMeta(manifest);
	const localLogo = opts.logoOverride ?? manifest.logoFile;
	let logoUpload;
	if (localLogo) {
		const logoPath = resolve(opts.cwd, localLogo);
		if (!existsSync(logoPath)) throw new CliError(`logo file not found: ${logoPath}`, 4);
		logoUpload = { source: localLogo };
	} else if (manifest.logoUrl) meta.logo_url = manifest.logoUrl;
	const screenshotSources = manifest.screenshots ?? [];
	const screenshotPlan = screenshotSources.map((src) => {
		const remote = isRemoteAsset(src);
		if (!remote && !existsSync(resolve(opts.cwd, src))) throw new CliError(`screenshot file not found: ${resolve(opts.cwd, src)}`, 4);
		return {
			source: src,
			remote
		};
	});
	const localScreenshots = screenshotPlan.filter((s) => !s.remote);
	const noop = Object.keys(meta).length === 0 && !logoUpload && screenshotSources.length === 0;
	if (opts.dryRun || noop) {
		if (screenshotSources.length > 0) meta.screenshots = screenshotSources;
		return {
			meta,
			logoUpload,
			screenshotUploads: localScreenshots.length > 0 ? localScreenshots.map((s) => ({ source: s.source })) : void 0,
			noop
		};
	}
	const screenshotUploads = [];
	const finalScreenshots = [];
	for (const item of screenshotPlan) {
		if (item.remote) {
			finalScreenshots.push(item.source);
			continue;
		}
		const shotPath = resolve(opts.cwd, item.source);
		const bytes = readFileSync(shotPath);
		const blob = new Blob([new Uint8Array(bytes)], { type: imageContentType(shotPath) });
		const res = await uploadAppScreenshot(client, appId, blob);
		finalScreenshots.push(res.screenshot_url);
		screenshotUploads.push({
			source: item.source,
			screenshot_url: res.screenshot_url
		});
	}
	if (finalScreenshots.length > 0) meta.screenshots = finalScreenshots;
	if (Object.keys(meta).length > 0) await patchApp(client, appId, meta);
	if (logoUpload) {
		const logoPath = resolve(opts.cwd, logoUpload.source);
		const bytes = readFileSync(logoPath);
		const blob = new Blob([new Uint8Array(bytes)], { type: imageContentType(logoPath) });
		const res = await uploadAppLogo(client, appId, blob);
		logoUpload.logo_url = res.logo_url;
	}
	return {
		meta,
		logoUpload,
		screenshotUploads: screenshotUploads.length > 0 ? screenshotUploads : void 0,
		noop
	};
}

//#endregion
export { syncAppListingMeta };
