// @anna-ai/cli/dist/publish-GpvkaBIx.js
import "./credentials-BTv2IfUZ.js";
import "./apps-DlVsVHQ5.js";
import { CliError } from "./client-D-_z1ALk.js";
import { withErrorHandling } from "./_lifecycle-shared-CUK2CE76.js";
import "./bundled-executas-_H1A9M50.js";
import "./executas-BhFed1B-.js";
import "./executa-cache-Do1D7xC0.js";
import { runExecutaPublish } from "./executa-publish-Ah6J7jYd.js";
import "./manifest-DOMrcdCK.js";
import "./app-cache-BqnTkxUw.js";
import "./listing-meta-BlY7XMrw.js";
import "./app-bundle-upload-CqT9KUOl.js";
import { runAppsPublish } from "./apps-publish-BKqVsS1W.js";
import { resolve } from "node:path";
import { existsSync } from "node:fs";
import { red, yellow } from "kleur/colors";

//#region src/publish/detector.ts
function detectProjectKind(cwd) {
	const hasManifest = existsSync(resolve(cwd, "manifest.json"));
	const hasExecuta = existsSync(resolve(cwd, "executa.json"));
	if (hasManifest && hasExecuta) return "ambiguous";
	if (hasManifest) return "app";
	if (hasExecuta) return "executa";
	return "unknown";
}

//#endregion
//#region src/commands/publish.ts
async function runTopLevelPublish(opts) {
	return withErrorHandling(async () => {
		const cwd = resolve(opts.cwd ?? process.cwd());
		const kind = detectProjectKind(cwd);
		if (kind === "ambiguous") throw new CliError(`cannot auto-detect: ${cwd} has BOTH manifest.json and executa.json. Run \`anna-app apps publish\` or \`anna-app executa publish\` explicitly.`, 4);
		if (kind === "unknown") throw new CliError(`no manifest.json or executa.json found in ${cwd}`, 4);
		if (kind === "app") return runAppsPublish({
			cwd,
			bump: opts.bump,
			noWrite: opts.noWrite,
			dryRun: opts.dryRun,
			account: opts.account,
			json: opts.json,
			noBundle: opts.noBundle,
			bundleDir: opts.bundleDir
		});
		return runExecutaPublish({
			cwd,
			bump: opts.bump,
			noWrite: opts.noWrite,
			publish: opts.publish,
			dryRun: opts.dryRun,
			account: opts.account,
			json: opts.json
		});
	});
}

//#endregion
export { runTopLevelPublish };
